core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/function/Function.java

```diff
@@ -32,11 +32,12 @@ public interface Function {
 	 * @param args         the function input arguments.
 	 * @return the function result value.
 	 * @throws ValueExprEvaluationException
-	 * @deprecated since 3.3.0. Use {@link #evaluate(TripleSource, Value...)} instead. A reference to a ValueFactory can
-	 *             be retrieved using {@link TripleSource#getValueFactory()} if needed.
+	 * @deprecated since 3.3.0. Use {@link #evaluate(TripleSource, Value...)} instead.
 	 */
 	@Deprecated
-	Value evaluate(ValueFactory valueFactory, Value... args) throws ValueExprEvaluationException;
+	default Value evaluate(ValueFactory valueFactory, Value... args) throws ValueExprEvaluationException {
+		throw new UnsupportedOperationException("use evaluate(TripleSource, Value...) instead");
+	}
```

Comment:
Should this maybe be targeted at 4.1.0?

Reply:
I'm a bit on the fence on that. I originally only wanted to change the docs and example, but I discovered this while doing so. What this change allows is a cleaner implementation where you don't _have_ to implement the deprecated method anymore (but still can if you want, so it's backward compatible). 

I think it's defensible to do this as a bug fix and put it in 4.0.1. Especially since we aim for a quick 4.0.1 release anyway, hopefully. The only thing I don't know for sure is if this is binary compatible.

This reminds me we should reenable the API compatibility checks for the main branch, now that 4.0 is out the door.

Reply:
Heh. Just ran a check on this, and it's not binary compatible. Which according to the japicmp plugin would make this a breaking change, needing a major version.